Implemented the Mamba State Space model

taddyb opened the pull request:
# Added:
- Implemented the [Mamba](https://github.com/state-spaces/mamba/tree/main) SSM within NH
  - NOTE: within Mamba there is a transition `nn.Linear` layer that changes the dimensions from the embedding layer to the hidden size. This is per Mamba's requirements as it expects inputs to be of the model's hidden size (read Mamba's README for more information)
```py
self.transition_layer = nn.Linear(self.embedding_net.output_size, self.cfg.hidden_size)
```
- made changes to the `Config.py` file to set the internal hyperparameters within Mamba
  - added: `d_state`, `d_conv`, and `expand`


# Tests:
- We ran a full 30 epoch validation testing and was able to generate mid performance (~0.4 NSE). Excited to see what experiments people run with this. 

![image](https://github.com/neuralhydrology/neuralhydrology/assets/16233925/79540c31-240b-43f2-b9ab-bcb08d4fcb50)

taddyb commented:
Note: there still needs to be a change to the requirements.txt, or environment.yml, file to include the following packages:
- causal-conv1d>=1.1.0
- mamba-ssm

<img width="728" alt="image" src="https://github.com/neuralhydrology/neuralhydrology/assets/16233925/d3c34f53-7262-4a80-b24f-f68668b39e46">

gauchm commented on neuralhydrology/utils/config.py:
```diff
@@ -311,6 +311,14 @@ def continue_from_epoch(self) -> int:
         return self._as_default_dict(self._cfg.get("custom_normalization", {}))
 
+    @property
+    def d_conv(self) -> dict:
```
type seems to be int, not dict

gauchm commented on neuralhydrology/utils/config.py:
```diff
@@ -311,6 +311,14 @@ def continue_from_epoch(self) -> int:
+        return self._cfg.get("d_conv", 4)
+
+    @property
+    def d_state(self) -> dict:
```
int

gauchm commented on neuralhydrology/utils/config.py:
```diff
@@ -370,6 +378,10 @@ def evolving_attributes(self) -> List[str]:
             return []
 
+    @property
+    def expand(self) -> dict:
```
int

gauchm commented on neuralhydrology/modelzoo/cudamamba.py:
```diff
@@ -0,0 +1,83 @@
+    To control the initial forget gate bias, use the config argument `initial_forget_bias`. Often it is useful to set
+    this value to a positive value at the start of the model training, to keep the forget gate closed and to facilitate
+    the gradient flow.
+    The `CudaMamba` class only supports single-timescale predictions. Use `???` to train a model and get
```
Is the ??? some artifact? Can we remove this?

gauchm reviewed (commented):
Thanks a lot, this is great!

Some general comments:
- I think I would like to make mamba and causal-conv1d optional dependencies, e.g. with a try-except around the import. The conda installation is already complicated enough and takes forever, so I fear that adding more deps will make it even worse. Not sure what we'll do with the environment files -- maybe keep as-is and tell the user in the exception how to install the missing dependencies?
- I'd vote for renaming the new config args to something like mamba_* to make it clear that they are used for mamba.
- We'll need some more documentation on the new model and its config arguments. For a recent example, see [here](https://github.com/neuralhydrology/neuralhydrology/commit/39fd3a8dec4e084bc5255ea88ac544a0c3ad56f8#diff-d24f8212d1d11307f7709e981cb8165ef1038e5fa2c2677a7ed3477c4771b016).

taddyb commented:
@gauchm Thanks for taking the time to review my changes. I implemented the following changes:
- Renamed `CudaMamba` to `Mamba`
- Added a try/except ModuleNotFoundError to the import of mamba_ssm
  - Tested the Try/except error statement
- Changed the new config entries to have `mamba_` in front of their naming 
- Changed Type hinting in config entries to `int`
- Added documentation to `config.rst` and `models.rst`
- Fixed Docstring `???` 

Hopefully everything passes 👍

gauchm commented:
Thanks for the changes. The import needs a small modification, and there is one more thing to do in the docs (sorry for not mentioning earlier):

Create a file like this one: https://github.com/neuralhydrology/neuralhydrology/blob/master/docs/source/api/neuralhydrology.modelzoo.gru.rst for Mamba and add it here: https://github.com/neuralhydrology/neuralhydrology/blob/master/docs/source/api/neuralhydrology.modelzoo.rst

taddyb commented:
@gauchm great spot of the global exception. Made the necessary changes and doc additions.

gauchm commented on neuralhydrology/modelzoo/__init__.py:
```diff
@@ -19,7 +20,8 @@
 SINGLE_FREQ_MODELS = [
-    "cudalstm", 
+    "cudalstm",
+    "cudamamba",
```
```suggestion
    "mamba",
```

gauchm reviewed (commented):
Thanks @taddyb! Looks good to me, except for a tiny mistake I failed to spot earlier.
Other than that, I think we're done.

taddyb commented:
Great! Just pushed the last changes 👍 @gauchm

kratzert commented:
Hi @taddyb, I am currently with Martin in Zurich an we do a focus week for some project and I haven't yet found time to look into this. I mostly trust Martin that he carefully reviewed your CL but if it is okay for you, I would wait until next week so that I can also take a look at it before we merge it. And +1 for doing a squash commit at the end. Thanks again for opening this PR and contributing the Mamba model.

kratzert commented on docs/source/usage/config.rst:
```diff
@@ -225,6 +225,14 @@ These are used if ``model == mtslstm``.
    current input timescale. In both cases, ``transfer_mtslstm_states``
    can be used to configure hidden and cell state transfer.
 
+Mamba settings
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
Mamba settings
~~~~~~~~~~~~~~
```
I think this has to be exactly the same lengths as the section title

kratzert commented on docs/source/api/neuralhydrology.modelzoo.mamba.rst:
```diff
@@ -0,0 +1,7 @@
+Mamba
+===
```
```suggestion
Mamba
=====
```
I think this has to be exactly the same lengths as the section title

taddyb commented:
Thanks for the review @kratzert. Just made the changes requested 👍

kratzert reviewed (approved):
Thanks, looks all good to me. I will go ahead and squash merge the PR into master. Thanks a lot for everything.
